mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ExecutorStateManagerImpl.java

```diff
@@ -284,57 +295,45 @@ public Optional<BestFit> findBestFit(TaskExecutorBatchAssignmentRequest request)
+    private Optional<Map<TaskExecutorID, TaskExecutorState>> findBestFitFor(TaskExecutorBatchAssignmentRequest request, SchedulingConstraints schedulingConstraints, Integer numWorkers, BestFit currentBestFit) {
         // only allow allocation in the lowest CPU cores matching group.
-        SortedMap<Double, NavigableSet<TaskExecutorHolder>> targetMap =
-            this.executorByCores.tailMap(machineDefinition.getCpuCores());
+        Optional<TaskExecutorGroupKey> bestFitTeGroupKey = findBestFitTeGroupKey(schedulingConstraints);
 
-        if (targetMap.isEmpty()) {
-            log.warn("Cannot find any executor for request: {}", request);
+        if (!bestFitTeGroupKey.isPresent()) {
+            log.warn("Cannot find any matching sku for request: {}", request);
             return Optional.empty();
         }
-        Double targetCoreCount = targetMap.firstKey();
-        log.debug("Applying assignmentReq: {} to {} cores.", request, targetCoreCount);
-
-        Double requestedCoreCount = machineDefinition.getCpuCores();
-        if (Math.abs(targetCoreCount - requestedCoreCount) > 1E-10) {
-            // this mismatch should not happen in production and indicates TE registration/spec problem.
-            log.warn("Requested core count mismatched. requested: {}, found: {} for {}", requestedCoreCount,
-                targetCoreCount,
-                request);
-        }
 
-        if (this.executorByCores.get(targetCoreCount).isEmpty()) {
-            log.warn("No available TE found for core count: {}, request: {}", targetCoreCount, request);
+        log.info("Applying assignment request: {} to constraints {}.", request, bestFitTeGroupKey);
+        if (!this.executorByGroupKey.containsKey(bestFitTeGroupKey.get())) {
+            log.warn("No available TE found for constraints: {}, request: {}", bestFitTeGroupKey.get(), request);
             return Optional.empty();
         }
 
         return Optional.of(
-            this.executorByCores.get(targetCoreCount)
+            this.executorByGroupKey.get(bestFitTeGroupKey.get())
                 .descendingSet()
                 .stream()
                 .filter(teHolder -> {
                     if (!this.taskExecutorStateMap.containsKey(teHolder.getId())) {
                         return false;
                     }
-
                     if (currentBestFit.contains(teHolder.getId())) {
                         return false;
                     }
-
                     TaskExecutorState st = this.taskExecutorStateMap.get(teHolder.getId());
                     return st.isAvailable() &&
                         st.getRegistration() != null &&
-                        st.getRegistration().getMachineDefinition().canFit(machineDefinition);
+                        st.getRegistration().getMachineDefinition().canFit(schedulingConstraints.getMachineDefinition());
```

Comment:
Do we need this check since the find best group already checks for the fitness? Can we just return the first TE in a group that's not assigned?

Reply:
that's fair, I think this check is now redundant.